src/ast/passes/codegen_llvm.cpp

```diff
@@ -3771,28 +3771,41 @@ void CodegenLLVM::generate_maps(const RequiredResources &required_resources,
                       1,
                       MapKey({ CreateInt(loss_cnt_key_size) }),
                       CreateInt(loss_cnt_val_size));
-
-  if (required_resources.max_fmtstring_args_size > 0) {
-    createMapDefinition(to_string(MapType::FmtStringArgs),
-                        libbpf::BPF_MAP_TYPE_PERCPU_ARRAY,
-                        1,
-                        MapKey({ CreateInt32() }),
-                        CreateArray(required_resources.max_fmtstring_args_size,
-                                    CreateInt8()));
-  }
 }
 
 void CodegenLLVM::generate_global_vars(const RequiredResources &resources)
 {
   for (const auto &name : resources.needed_global_vars) {
-    auto var = llvm::dyn_cast<GlobalVariable>(
-        module_->getOrInsertGlobal(name, b_.getInt64Ty()));
-    var->setInitializer(b_.getInt64(1));
-    var->setConstant(true);
-    var->setSection(bpftrace::globalvars::SECTION_NAME);
-    var->setExternallyInitialized(true);
-    var->setDSOLocal(true);
-    var->addDebugInfo(debug_.createGlobalInt64(name));
+    if (name == globalvars::NUM_CPUS) {
+      auto var = llvm::dyn_cast<GlobalVariable>(
+          module_->getOrInsertGlobal(name, b_.getInt64Ty()));
+      var->setInitializer(b_.getInt64(1));
+      var->setConstant(true);
+      var->setSection(bpftrace::globalvars::RO_SECTION_NAME);
+      var->setExternallyInitialized(true);
+      var->setDSOLocal(true);
+      var->addDebugInfo(debug_.createGlobalInt64(name));
+    } else if (name == globalvars::FMT_STRINGS_BUFFER) {
+      assert(resources.max_fmtstring_args_size > 0);
+      // Assume 1 CPU to ensure codegen tests pass across multiple machines and
+      // so LLVM doesn't incorrectly optimize code (which happens if array size
+      // = 1)
+      auto array_type = CreateArray(resources.max_fmtstring_args_size,
+                                    CreateInt8());
```

Comment:
Seems reasonable. Typically how we do it in other places is there's some member function in a struct that returns the value. And for codegen we'll override the implementation.

But probably simpler to do it like this.

Reply:
Ooh that's a good idea, that's a little cleaner, I'll try that if its easy enough

Reply:
Sorry I ended up not trying out (spent a lot of time refactoring globalvars), can you send me a code ref if you have one?

Reply:
Oh, that's not what I meant but also a good idea. I was more referring to ensuring the codegen unit tests are stable. ie we wouldn't want the test output to change depending on how many CPUs your laptop has.

Here you're essentially setting a random (but stable) size in codegen. Which ensures the `.ll` output is stable. At runtime it'll be resized to the real size. Which I think is probably the simplest way to do it.

In other places we use mocking to ensure stability. Eg: https://github.com/bpftrace/bpftrace/blob/3cd6d50e3597dfd826583453c318c39002044572/tests/codegen/array_integer_equal_comparison.cpp#L20-L30

But perhaps overkill for this

Reply:
Ah I see, agreed, this approach is simpler so I'll go with this for now!